### PR TITLE
EquipmentChangeEvent for 1.11

### PR DESCRIPTION
**Pull request by TechnicianLP**
Added a EquipmentChangeEvent into EntityLivingBase (same as #3285 but for 1.11)
Fell free to make suggestions!

### CODE REVIEW

**Comment by TechnicianLP** [comment]
welp will have to edit my name there ...

**Comment by TechnicianLP** [comment]
i guess it works now ...

**Comment by mezz** [comment]
Looks good, and should help a bit with performance because people can switch to use this instead of much broader events.

Please create a Forge example mod to show that this works with confidence.
If you need help getting the example mods running, see [cpw's video tutorial](https://www.youtube.com/watch?v=pLWQk6ed56Q).

**Comment by LatvianModder** [comment]
You got captial H in title :P

**Comment by TechnicianLP** [comment]
@mezz Examplemod added
@LatvianModder that definitly is on purpose

**Comment by mezz** [comment]
Tested and looks good to me.
```
[01:58:01] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in FEET from 1xtile.air@0 to 1xitem.bootsIron@0
[01:58:03] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in FEET from 1xitem.bootsIron@0 to 1xtile.air@0
[01:58:08] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in CHEST from 1xtile.air@0 to 1xitem.chestplateGold@0
[01:58:10] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in CHEST from 1xitem.chestplateGold@0 to 1xtile.air@0
[01:58:12] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in CHEST from 1xtile.air@0 to 1xitem.chestplateIron@0
[01:58:12] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in CHEST from 1xitem.chestplateIron@0 to 1xtile.air@0
[01:58:14] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in FEET from 1xtile.air@0 to 1xitem.bootsCloth@0
[01:58:14] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in FEET from 1xitem.bootsCloth@0 to 1xtile.air@0
[01:58:15] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in LEGS from 1xtile.air@0 to 1xitem.leggingsChain@0
[01:58:17] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in LEGS from 1xitem.leggingsChain@0 to 1xtile.air@0
[01:58:18] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in CHEST from 1xtile.air@0 to 1xitem.chestplateChain@0
[01:58:19] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in CHEST from 1xitem.chestplateChain@0 to 1xtile.air@0
[01:58:20] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in MAINHAND from 1xtile.air@0 to 1xitem.helmetDiamond@0
[01:58:22] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in MAINHAND from 1xitem.helmetDiamond@0 to 1xtile.air@0
[01:58:22] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in OFFHAND from 1xtile.air@0 to 1xitem.helmetDiamond@0
[01:58:24] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in MAINHAND from 1xtile.air@0 to 1xitem.leggingsDiamond@0
[01:58:24] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in MAINHAND from 1xitem.leggingsDiamond@0 to 1xitem.helmetGold@0
[01:58:26] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in MAINHAND from 1xitem.helmetGold@0 to 1xitem.helmetDiamond@0
[01:58:26] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in OFFHAND from 1xitem.helmetDiamond@0 to 1xitem.helmetGold@0
[01:58:30] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in MAINHAND from 1xitem.helmetDiamond@0 to 1xtile.air@0
[01:58:32] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in MAINHAND from 1xtile.air@0 to 1xitem.helmetGold@0
[01:58:32] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in OFFHAND from 1xitem.helmetGold@0 to 1xtile.air@0
[01:58:33] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/147, l='New World', x=16.60, y=4.00, z=758.47] changed his Equipment in MAINHAND from 1xitem.helmetGold@0 to 1xtile.air@0
```

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch [hunk 1]
```diff
@@ -199,6 +199,15 @@
          super.func_70071_h_();
          this.func_184608_ct();
  
+@@ -2015,7 +2041,7 @@
+                 if (!ItemStack.func_77989_b(itemstack1, itemstack))
+                 {
+                     ((WorldServer)this.field_70170_p).func_73039_n().func_151247_a(this, new SPacketEntityEquipment(this.func_145782_y(), entityequipmentslot, itemstack1));
+-
```
Comment (mezz) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: I think you can reduce your patch by leaving this newline in

**Review by mezz — CHANGES_REQUESTED** [review]
one tiny change

**Comment by HenryLoenwind** [comment]
Is this also fired when joining a world or when the player object is cloned?

**Comment by TechnicianLP** [comment]
It definitly fires when joining the World as the Entity did not have Equipment before - it works for cloning the player as well

**Review by mezz — APPROVED** [review]

**Comment by mezz** [comment]
Tested these cases to verify, looks good to me.

Joining world while wearing a helmet:
```
[09:07:45] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/57691, l='Copy of EquipmentChangeEvent', x=-202.50, y=4.00, z=-320.50] changed his Equipment in HEAD from 1xtile.air@0 to 1xitem.helmetIron@0
```

Making a nether portal:
```
[09:08:41] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/57691, l='Copy of EquipmentChangeEvent', x=-202.50, y=4.00, z=-320.50] changed his Equipment in MAINHAND from 1xtile.air@0 to 1xtile.obsidian@0
[09:08:54] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/57691, l='Copy of EquipmentChangeEvent', x=-202.30, y=6.49, z=-319.66] changed his Equipment in MAINHAND from 1xtile.obsidian@0 to 1xitem.flintAndSteel@0
```

Entering the nether:
```
[09:09:06] [Server thread/INFO] [FML]: [Equipment-Change] EntityPigZombie['Zombie Pigman'/334280, l='Copy of EquipmentChangeEvent', x=-67.50, y=42.00, z=-52.50] changed his Equipment in MAINHAND from 1xtile.air@0 to 1xitem.swordGold@0
[09:09:06] [Server thread/INFO] [FML]: [Equipment-Change] EntityPigZombie['Zombie Pigman'/334281, l='Copy of EquipmentChangeEvent', x=-118.50, y=35.00, z=12.50] changed his Equipment in MAINHAND from 1xtile.air@0 to 1xitem.swordGold@0
[09:09:06] [Server thread/INFO] [FML]: [Equipment-Change] EntityPigZombie['Zombie Pigman'/334287, l='Copy of EquipmentChangeEvent', x=-33.50, y=44.00, z=-140.50] changed his Equipment in MAINHAND from 1xtile.air@0 to 1xitem.swordGold@0
[09:09:06] [Server thread/INFO] [FML]: [Equipment-Change] EntityPigZombie['Zombie Pigman'/334288, l='Copy of EquipmentChangeEvent', x=-36.50, y=44.00, z=-140.50] changed his Equipment in MAINHAND from 1xtile.air@0 to 1xitem.swordGold@0
... lots more pigmen
```

Respawning after defeating the ender dragon and going into the portal:
```
[09:28:13] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/445857, l='EquipmentChangeEvent', x=-247.50, y=70.00, z=-43.50] changed his Equipment in MAINHAND from 1xtile.air@0 to 1xitem.swordDiamond@0
[09:28:13] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/445857, l='EquipmentChangeEvent', x=-247.50, y=70.00, z=-43.50] changed his Equipment in FEET from 1xtile.air@0 to 1xitem.bootsDiamond@0
[09:28:13] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/445857, l='EquipmentChangeEvent', x=-247.50, y=70.00, z=-43.50] changed his Equipment in LEGS from 1xtile.air@0 to 1xitem.leggingsDiamond@0
[09:28:13] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/445857, l='EquipmentChangeEvent', x=-247.50, y=70.00, z=-43.50] changed his Equipment in CHEST from 1xtile.air@0 to 1xitem.chestplateDiamond@0
[09:28:13] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/445857, l='EquipmentChangeEvent', x=-247.50, y=70.00, z=-43.50] changed his Equipment in HEAD from 1xtile.air@0 to 1xitem.helmetDiamond@0
```

Using a dispenser to add items to the player inventory:
```
[16:41:09] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/23, l='New World', x=-759.54, y=4.75, z=603.53] changed his Equipment in FEET from 1xtile.air@0 to 1xitem.bootsDiamond@0
```

Using a shovel to dig until it breaks:
```
[16:41:28] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/23, l='New World', x=-760.60, y=4.00, z=600.68] changed his Equipment in MAINHAND from 1xitem.shovelGold@0 to 1xitem.shovelGold@1
[16:41:28] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/23, l='New World', x=-760.60, y=4.00, z=600.68] changed his Equipment in MAINHAND from 1xitem.shovelGold@1 to 1xitem.shovelGold@2
[16:41:28] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/23, l='New World', x=-760.35, y=4.00, z=600.74] changed his Equipment in MAINHAND from 1xitem.shovelGold@2 to 1xitem.shovelGold@3
... etc ...
[16:41:41] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/23, l='New World', x=-743.88, y=2.00, z=601.51] changed his Equipment in MAINHAND from 1xitem.shovelGold@30 to 1xitem.shovelGold@31
[16:41:41] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/23, l='New World', x=-743.78, y=2.00, z=601.52] changed his Equipment in MAINHAND from 1xitem.shovelGold@31 to 1xitem.shovelGold@32
[16:41:42] [Server thread/INFO] [FML]: [Equipment-Change] EntityPlayerMP['mezz'/23, l='New World', x=-742.84, y=2.00, z=601.52] changed his Equipment in MAINHAND from 1xitem.shovelGold@32 to 0xtile.air@0
```

**Comment by mezz** [comment]
@TechnicianLP did you edit the patch by hand? Never do that, it does not work. 
Please fix the code and generate it correctly with gradle genPatches.

**Comment by TechnicianLP** [comment]
Seems like there has been some Problem while copying the new Patch ... will fix it as soon as i can

**Comment by TechnicianLP** [comment]
@mezz Patch is fixed now - i dont know which part failed at my end there ...

**Comment by mezz** [comment]
Ok tested and looks good.

**Pull request merged**